p2p/types/src/network_address.rs

```diff
@@ -0,0 +1,452 @@
+
+    async fn resolve(
+        &self,
+    ) -> Result<impl Iterator<Item = Self::ResolvedAddress> + '_, NameResolutionError> {
+        match self {
+            Self::Ip(ip) => Ok(Either::Left(std::iter::once(*ip))),
+            Self::Domain(domain_name) => Ok(Either::Right(domain_name.resolve().await?)),
+        }
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = NetworkAddressParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(ip) = s.parse::<IpAddr>() {
+            Ok(NetworkAddress::Ip(ip))
+        } else {
+            Ok(NetworkAddress::Domain(s.parse::<DomainName>()?))
+        }
+    }
+}
+
+impl Display for NetworkAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Ip(addr) => {
+                write!(f, "{addr}")
+            }
+            Self::Domain(domain_name) => {
+                write!(f, "{domain_name}")
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, DeserializeFromStr, SerializeDisplay)]
+pub struct NetworkAddressWithOptionalPort {
+    pub address: NetworkAddress,
+    pub port: Option<u16>,
+}
+
+impl ResolvableName for NetworkAddressWithOptionalPort {
+    type ResolvedAddress = IpOrSocketAddress;
+
+    async fn resolve(
+        &self,
+    ) -> Result<impl Iterator<Item = IpOrSocketAddress> + '_, NameResolutionError> {
+        Ok(self.address.resolve().await?.map(|ip_addr| {
+            if let Some(port) = self.port {
+                IpOrSocketAddress::new_socket_address(SocketAddr::new(ip_addr, port))
+            } else {
+                IpOrSocketAddress::new_ip(ip_addr)
+            }
+        }))
+    }
+}
+
+impl FromStr for NetworkAddressWithOptionalPort {
+    type Err = NetworkAddressParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // With ipv6 addresses, finding the port separator is not super trivial, so we want
+        // to delegate that work to SocketAddr's FromStr. But it won't consider something like
+        // '1:2:3:4:5:6:7:8' as a valid ipv6 address, so try IpAddr's FromStr first.
+
+        if let Ok(ip_addr) = s.parse::<IpAddr>() {
+            return Ok(Self {
+                address: NetworkAddress::Ip(ip_addr),
+                port: None,
+            });
+        }
+
+        if let Ok(socket_addr) = s.parse::<SocketAddr>() {
+            return Ok(Self {
+                address: NetworkAddress::Ip(socket_addr.ip()),
+                port: Some(socket_addr.port()),
+            });
+        }
+
+        let separator = ':';
+        assert!(separator.is_ascii());
+        let separator_len = 1;
+
+        // Use rfind instead of find to produce better errors; e.g. for something like
+        // '[1:2:3:4:5:6:7:8]:foo' we'd prefer the error 'BadPortNumber("foo")' instead of
+        // 'BadPortNumber("2:3:4:5:6:7:8]:foo")'
+        let separator_pos = s.rfind(separator);
```

Comment:
What if there's no port number specified and it's only an IPv6 address?

Reply:
> What if there's no port number specified and it's only an IPv6 address?

Then it should be specified without brackets. There is a test for it.